text/0000-std-rpc-api.md

```diff
@@ -0,0 +1,433 @@
+
+# Motivation
+[motivation]: #motivation
+
+Achieve better performance, unify & standardize RPC API, make RPC API extensible & future-proof and reduce boilerplate code.
+
+Also a simple internal performance benchmark came with these results:
+
+    Local RPC Send Performance Test (524280 calls)
+    Convenience -> 4361.0433ms
+    Performance -> 2607.7266ms
+    Standard -> 2205.043ms
+    Local RPC Recv Performance Test (524280 calls)
+    Convenience -> 9626.3536ms
+    Performance -> 5923.7964ms
+    Standard -> 1935.296ms
+
+We're looking at **up to 5.24x better performance** (compared to convenience API vs standard API receive performance).
+
+Beyond that, on an internal project, we saw roughly **%80 less boilerplate code** compared to performance RPCs and standard RPCs which is also a huge win.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Multiplayer framework provides 2 main network constructs ([ServerRPC](#serverrpc) and [ClientRPC](#clientrpc)) to execute logic on either server-side or client-side. This concept often called as [Remote Procedure Call (RPC)](https://en.wikipedia.org/wiki/Remote_procedure_call) and has wide adoption across the industry.
+
+## RPC Methods
+
+A framework user (Unity developer) can declare multiple RPCs under a `NetworkBehaviour` and inbound/outbound RPC calls will be replicated as a part of its replication in a network frame.
+
+### ServerRPC
+
+A `ServerRPC` can be invoked by a client to be executed on the server.
+
+Developer can declare a `ServerRPC` by marking a method with `[ServerRPC]` attribute and making sure to have `ServerRPC` postfix in the method name:
+
+```cs
+[ServerRPC]
+void PingServerRPC(int framekey) { /* ... */ }
+```
+
+Developer can invoke a `ServerRPC` by making a direct function call with parameters:
+
+```cs
+void Update()
+{
+    if (Input.GetKeyDown(KeyCode.P))
+    {
+        PingServerRPC(Time.frameCount); // Client -> Server
+    }
+}
+```
+
+Marking method with `[ServerRPC]` and putting `ServerRPC` postfix to the method name are required:
+
+```cs
+// Invalid, missing 'ServerRPC' postfix in the method name
+[ServerRPC]
+void Ping(int framekey) { /* ... */ }
+
+// Invalid, missing [ServerRPC] attribute
+void PingServerRPC(int framekey) { /* ... */ }
+```
+
+`[ServerRPC]` attribute and matching `...ServerRPC` postfix in the method name are there to make it crystal clear for RPC call sites to know when they are executing an RPC, it will be replicated and executed on the server-side, without necessarily jumping into original RPC method declaration to find out if it was an RPC, if so whether it is a ServerRPC or ClientRPC:
+
+```cs
+Ping(framekey); // Is this an RPC call?
+
+PingRPC(framekey); // Is this a ServerRPC call or ClientRPC call?
+
+PingServerRPC(framekey); // This is clearly a ServerRPC call
+```
```

Comment:
I like them being case-insensitive so that they can adapt to different coding styles (I prefer RPC, Rider prefers Rpc and refuses to accept my configuration changes, it seems).

Would it also be possible to allow both, postfixes and prefixes? So, one could write Ping_ServerRPC_ but also _ServerRPC_Ping? At the moment, I feel like the postfix will usually be preferable but there are probably cases where intellisense grouping by method type may be preferable. Postfix is better for "what methods are available, which type of method do I need right now", prefix is better for "what server rpcs does this class offer, ok, I need this specific method".

I think it makes sense to enforce the naming convention - but when it's enforced, it should adapt to different approaches, IMHO.

Reply:
One more thing regarding the naming convention: ServerRPC and ClientRPC may be too verbose for some people. So allowing SRPC and CRPC and *maybe* SrvRPC and CliRPC in addition may be appreciated by those. It may not matter much with Ping but sometimes the best name for a method might be DoThatAwesomeThing. In fact, even with Ping, a shorter postfix/prefix may actually improve readability quite a bit.

Reply:
Note that something similar is already in place to help with Unity Messages. If the method signature is wrong there is a helpful, yet magic, error message in the Unity Editor.
![image](https://user-images.githubusercontent.com/2233980/101623277-7f0b3880-3a18-11eb-9d0d-ef3fc279e976.png)

Reply:
> The difference with this and other magic-names is that this is actually doing some magic under the hood - we're breaking the general convention of, "when you call a function, it will be invoked synchronously before returning control to the parent function". This is going to make comprehension a lot more difficult for users if these functions look like every other function at the call site. Without drilling down, you have no way of knowing that the function you're calling is actually an RPC that will be invoked asynchronously.
> I wouldn't say that the naming suffix requirement is an elegant solution, but it does at least improve call site readability. Now the user can build some expectations when reading code, "ok this behavior can only happen when I see this suffix". Even though it's a little awkward for the end user, I think we end up saving them a lot of pain if we do this.

This i think is very important to mention, and is something that trips up new users already. A few days ago on the Discord a user was confused about the fact that it doesnt execute immediately.

Reply:
Why do i both need an attribute and the naming convention to match? Isnt the attribute redundant?

Reply:
> `[ServerRpc]` attribute and matching `...ServerRpc` suffix in the method name are there to make it crystal clear for RPC call sites to know when they are executing an RPC, it will be replicated and executed on the server-side, without necessarily jumping into original RPC method declaration to find out if it was an RPC, if so whether it is a ServerRpc or ClientRpc:
> 
> ```cs
> Ping(framekey); // Is this an RPC call?
> 
> PingRpc(framekey); // Is this a ServerRpc call or ClientRpc call?
> 
> PingServerRpc(framekey); // This is clearly a ServerRpc call
> ```

Reply:
The more i think about it, the less i like a C# framework forcing naming conventions on us. Several reasons:

1. Although its a fine guideline, others can be proposed that are equally good/bad. Personally i started prefixing methods with `client` or `server`, that made most sense for me. I prefix because otherwise you get methods that are called `validateClientServerRpc`, vs `serverValidateClient`. I find the latter is clearer.
2. Some developers have code bases in their native language, where this wouldnt fit well.
3. It might make migrating from another networking stack more difficult.
4. Within C# this is very uncommon, besides the magical monobehaviour callback methods Unity has there is as far as i am aware no other mainstream framework that does it. There is no prior-art. 
5. There is no technical reason, the attribute is enough. A strong recommendation achieves nearly the same.

Reply:
@TJHeuvel — I think that's fair, enforcing name convention by hard compile errors seems a little bit harsh. Maybe we can stick with compile/build warnings instead of hard-errors to attract more attention from the developer.
_(I think I'm changing my mind around too here)._

Microsoft does something similar with Async APi where you're overwhelmingly encouraged to put `...Async` suffix to your async task methods but you're not forced to. Usually you see a warning message popping up on that method signature and method call sites advising you to use `...Async` suffix but not forcing it hard.

It all comes down to "is tooling will be enough?" and it's also fair to say Unity Diagnostics Messages and Roslyn Analzyers will be covering that part thoroughly.

(I can see debug time spent figuring out what that method didn't replicate or vice-versa even today but after all, you can only enforce so much)

Reply:
One alternative to this is to provide a project level MLAPI setting for the naming convention of methods. Two regex fields - "ServerRpcValidation" and "ClientRpcValidation". An empty string would indicate no validation, and otherwise the regex would be used in the compilation step to validate that RPC methods follow a naming convention specified by the user. If we wanted to we could default it to the convention we're providing, but then we're not boxing users into a convention they may not want.

Reply:
Cant users do that already with Roslyn analysers, or EditorConfig with custom rules?

Reply:
Yes they can, this alternative would just be to provide the validation functionality without having to build it yourself

Reply:
I would also prefer to leave this as a best practice, backed by IDE hints, Roslyn and the like.

Reply:
I won't veto the process over it, but for the record I would prefer to not force the suffix.  I guess it's the language designer part of me that feels that devs cement their mental models & understanding when they see the 'why' of each piece of syntax.  I think folks will be consufed on why you need both the attribute and the suffix - it will lead them to wonder if they are doing 2 disparate things each handled by one of the 2.

Reply:
I will create another RFC & PR that suggest relaxing suffix enforcement but also PR will be implementing Roslyn Analyzers so that it would be still a pleasant experience for developers to see warnings/hints suggesting a better naming convention. (My main argument was to prevent people from making silly mistakes and there is no Roslyn Analyzers implementation yet but that'd change very soon with Unity2020.2 🎉)